src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -131,18 +191,46 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
                                 )),
                             ))
                         } else {
-                            Err(ErrorKind::ResolutionFailure)
+                            Err(ErrorKind::Resolve(ResolutionFailure::NotAVariant(
+                                ty_res,
+                                variant_field_name,
+                            )))
                         }
                     }
-                    _ => Err(ErrorKind::ResolutionFailure),
+                    _ => unreachable!(),
                 }
             }
-            _ => Err(ErrorKind::ResolutionFailure),
+            // `variant_field` looks at 3 different path segments in a row.
+            // But `NoAssocItem` assumes there are only 2. Check to see if there's
+            // an intermediate segment that resolves.
+            _ => {
+                let intermediate_path = format!("{}::{}", path, variant_name);
+                // NOTE: we have to be careful here, because we're already in `resolve`.
+                // We know this doesn't recurse forever because we use a shorter path each time.
+                // NOTE: this uses `TypeNS` because nothing else has a valid path segment after
+                let kind = if let Some(intermediate) = self.check_full_res(
+                    TypeNS,
+                    &intermediate_path,
+                    Some(module_id),
+                    current_item,
+                    extra_fragment,
+                ) {
+                    ResolutionFailure::NoAssocItem(intermediate, variant_field_name)
+                } else {
+                    // Even with the shorter path, it didn't resolve, so say that.
+                    ResolutionFailure::NoAssocItem(ty_res, variant_name)
+                };
```

Comment:
Actually this would be a good place to look up the first path segment that didn't resolve - just make this a loop instead of doing it once:
```rust
let err = loop {
     if let Some(res) = self.check_full_res(intermediate_path) {
        break res;
    }
    intermediate_path = match intermediate_path.rsplitn(2, "::").nth(1) {
       None => break ResolutionFailure::NoAssocItem(ty_res, variant_name),
       Some(inner) => inner,
    };
}
```

Reply:
It shouldn't go here, but slightly above where `ty_res` failed to resolve. I implemented this :)